privilege/privileges/privileges.go

```diff
@@ -0,0 +1,254 @@
+// Copyright 2015 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package privileges
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/model"
+	"github.com/pingcap/tidb/mysql"
+	"github.com/pingcap/tidb/privilege"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/sqlexec"
+)
+
+var _ privilege.Checker = (*PrivilegeCheck)(nil)
+
+type userPrivilege struct {
+	GlobalPrivs map[mysql.PrivilegeType]bool
+	DBPrivs     map[string]map[mysql.PrivilegeType]bool
+	TablePrivs  map[string]map[string]map[mysql.PrivilegeType]bool
+}
+
+// PrivilegeCheck implements privilege.Checker interface.
+// This is used to check privilege for the current user.
+type PrivilegeCheck struct {
+	username string
+	host     string
+	privs    *userPrivilege
+}
+
+// CheckDBPrivilege implements PrivilegeChecker.CheckDBPrivilege interface.
+func (p *PrivilegeCheck) CheckDBPrivilege(ctx context.Context, db *model.DBInfo, privilege mysql.PrivilegeType) (bool, error) {
+	if p.privs == nil {
+		// Lazy load
+		err := p.loadPrivileges(ctx)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+	}
+	// Check global scope privileges.
+	_, ok := p.privs.GlobalPrivs[privilege]
+	if ok {
+		return true, nil
+	}
+	// Check db scope privileges.
+	dbp, ok := p.privs.DBPrivs[db.Name.O]
+	if !ok {
+		return false, nil
+	}
+	_, ok = dbp[privilege]
+	return ok, nil
+}
+
+// CheckTablePrivilege implements PrivilegeChecker.CheckTablePrivilege interface.
+func (p *PrivilegeCheck) CheckTablePrivilege(ctx context.Context, db *model.DBInfo, tbl *model.TableInfo, privilege mysql.PrivilegeType) (bool, error) {
+	if p.privs == nil {
+		// Lazy load
+		err := p.loadPrivileges(ctx)
+		if err != nil {
+			return false, errors.Trace(err)
+		}
+	}
+	// Check global scope privileges.
+	_, ok := p.privs.GlobalPrivs[privilege]
+	if ok {
+		return true, nil
+	}
+	// Check db scope privileges.
+	dbp, ok := p.privs.DBPrivs[db.Name.O]
+	if ok {
+		_, ok = dbp[privilege]
+		if ok {
+			return true, nil
+		}
+	}
+	// Check table scope privileges.
+	tblp, ok := p.privs.TablePrivs[db.Name.O]
+	if !ok {
+		return false, nil
+	}
+	_, ok = tblp[tbl.Name.O]
+	return ok, nil
+}
+
+func (p *PrivilegeCheck) loadGlobalPrivileges(ctx context.Context) error {
+	sql := fmt.Sprintf(`SELECT * FROM %s.%s WHERE User="%s" AND (Host="%s" OR Host="%%");`, mysql.SystemDB, mysql.UserTable, p.username, p.host)
```

Comment:
Host="%%" means all host is ok. User has privileges granted to user@host and user@%.